### PR TITLE
Adds Domains Dashboard to My Site (and a Feature Flag for it)

### LINKED ISSUE #17067: [iOS] Domains: Add feature flag
Adds a feature flag for the new Domains Dashboard and related features

### LINKED ISSUE #17068: [iOS] Domains: Add entry point in My Site
Adds a new Entry in My Site that will be used to access the Domains Dashboard.

### LINKED ISSUE #17069: iOS - Domains Dashboard (initial layout)
Adds a Domains Dashboard screen, that will contain:
- Primary Site address
- An action to purchase a new domain
- An action to redeem a free domain with a paid plan, if applicable (that is: if the site has a paid plan and has not redeemed a domain yet)

### PR DESCRIPTION
**Pull request by Gio2018**
Fixes #17067, #17068, #17069

This PR adds a Feature Flag for Domains, and an entry point and initial implementation for the Domains Dashboard, accessible from My Site

To test:
Pre requisites:
- Use a store sandbox token to avoid actual domain purchases
- Have at least one site with a free plan
- Have at least one site with a paid plan that has not redeemed the free domain yet

Testing steps:
Test on iPhone and iPad
- Build/run and go to My Site
- Make sure the "Domains" entry appears in the "Configure" section (see screenshot)
- Select a site with a paid plan that has not yet redeemed the domain yet
- Tap on "Domains" and make sure the Domains Dashboard appears and looks like in the screenshot below (again, the duplicate "Register a domain" entry will be removed when it's final)
- Tap on "Register a free domain" and complete the flow: make sure the domain registration completes successfully, and that you see the registration epilogue at the end.
- Make sure that after dismissing the epilogue screen, you are redirected to the dashboard, that you don't see the "Register a free domain" section anymore and that the primary domain has been updated accordingly
- Go back to my site and select a free plan
- Tap on "Domains" and make sure you don't see the "Register a free domain" section

<p align=center>

</p>

## Regression Notes
1. Potential unintended areas of impact
None: this buds a brand new view

2. What I did to test those areas of impact (or what existing automated tests I relied on)

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift [hunk 1]
```diff
@@ -96,6 +99,8 @@ extension FeatureFlag {
             return "Recommend App to Others"
         case .newCommentEdit:
             return "New Comment Edit"
+        case .domains:
+            return "Domains selling"
```
Comment (diegoreymendez) [inline WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift]: I believe this is internally visible only, and if that's the case this is trivial and can be ignored.

Isn't the feature "Domain Purchases" from the PoV of the user?

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/DomainSuggestionViewControllerWrapper.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import SwiftUI
+import UIKit
+
+/// Makes RegisterDomainSuggestionsViewController available to SwiftUI
+final class DomainSuggestionViewControllerWrapper: UIViewControllerRepresentable {
+
+    let blog: Blog
+
+    weak var presentingController: RegisterDomainSuggestionsViewController?
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/DomainSuggestionViewControllerWrapper.swift]: Both properties could be private.

File: WordPress/Classes/ViewRelated/Blog/Blog Details/BlogDetailsViewController.m [hunk 3]
```diff
@@ -981,6 +981,19 @@ - (BlogDetailsSection *)configurationSectionViewModel
 
     [rows addObject:row];
 
+    if ([Feature enabled:FeatureFlagDomains]) {
+        BlogDetailsRow *domainsRow = [[BlogDetailsRow alloc] initWithTitle:NSLocalizedString(@"Domains", @"Noun. Title. Links to the Domains screen.")
+                                                         identifier:BlogDetailsSettingsCellIdentifier
+                                            accessibilityIdentifier:@"Domains Row"
+                                                                     image:[UIImage gridiconOfType:GridiconTypeDomains]
+                                                           callback:^{
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Blog/Blog Details/BlogDetailsViewController.m]: Nitpicky: the indentation seems to be off.

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift [hunk 4]
```diff
@@ -139,3 +141,36 @@ extension RegisterDomainSuggestionsViewController: NUXButtonViewControllerDelega
         self.navigationController?.pushViewController(controller, animated: true)
     }
 }
+
+// MARK: - DomainCreditRedemptionSuccessViewControllerDelegate
+
+extension RegisterDomainSuggestionsViewController: DomainCreditRedemptionSuccessViewControllerDelegate {
+
+    /// Handles the dismission of the epilogue screen at the end of a successful domain registration, along with popping to the domains dashboard
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift]: "dismission" > "dismissal"

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/DomainSuggestionViewControllerWrapper.swift (hunk 2 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/DomainSuggestionViewControllerWrapper.swift]: Additionally, I think `presentingViewController` should probably be `presentedViewController`.

I'm using this as my reference for the naming suggestion: https://developer.apple.com/documentation/uikit/uiviewcontroller/1621430-presentingviewcontroller

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift [hunk 5]
```diff
@@ -139,3 +141,36 @@ extension RegisterDomainSuggestionsViewController: NUXButtonViewControllerDelega
...
+            }) ?? self?.navigationController?.topViewController {
+                self?.navigationController?.popToViewController(popController, animated: true)
+            }
+
+            guard let email = self?.accountEmail() else {
+                return
+            }
+            let title = String(format: NSLocalizedString("Verify your email address - instructions sent to %@",
+                                                         comment: "Notice displayed after domain credit redemption success."), email)
+            ActionDispatcher.dispatch(NoticeAction.post(Notice(title: title)))
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift]: Just double checking with you: I imagine this isn't fully implemented yet, is that correct?

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift [hunk 6]
```diff
@@ -139,3 +141,36 @@ extension RegisterDomainSuggestionsViewController: NUXButtonViewControllerDelega
...
+
+    /// Handles the dismission of the epilogue screen at the end of a successful domain registration, along with popping to the domains dashboard
+    func continueButtonPressed() {
+
+        dismiss(animated: true) { [weak self] in
+            if let popController = self?.navigationController?.viewControllers.first(where: {
+                                                                                $0 is UIHostingController<DomainsDashboardView>
+            }) ?? self?.navigationController?.topViewController {
+                self?.navigationController?.popToViewController(popController, animated: true)
+            }
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift]: Any chance we can implement this without having to "know" there's a `UIHostingController<DomainsDashboardView>` before this VC?

This approach will work for us now but I fear it'll make this component harder to use in the future.

Maybe we can call a completion callback and let the previous VC handle its own dismissal logic?

**Review by diegoreymendez — CHANGES_REQUESTED** [review]
Hey @Gio2018.  The feature worked mostly fine for me up until the point where the primary domain is set.  Unfortunately for some reason that call fails for me in our testing environment.

I'll try to see if I can help moving that call to happen asynchronously so it doesn't block users with the registration if it fails, but for the time being I'm unable to complete the full flow.

I added several review comments.  Let me know if there's any doubt.

**Comment by diegoreymendez** [comment]
Just a final note: I managed to complete the flow.  Setting the primary domain fails if you don't have an upgrade in the sandbox store.

The reason I didn't realize this was the issue is because the free domain notice was coming up for me (probably since the App launches with the regular store, and loads the state for the site), and the purchase was succeeding because I had credits to spend in the sandbox store.  Weirdly enough the free-credits purchase succeeds using your credits, but setting the primary domain doesn't.

File: WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift (hunk 1 above)
Comment (Gio2018) [reply WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift]: Yes it is!! Thanks for spotting this

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/DomainSuggestionViewControllerWrapper.swift (hunk 2 above)
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/DomainSuggestionViewControllerWrapper.swift]: 👍

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift (hunk 5 above)
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift]: We can put a TODO here: this notice might need to change. This is what we are currently doing for the existing domain registration

**Comment by Gio2018** [comment]
> Just a final note: I managed to complete the flow. Setting the primary domain fails if you don't have an upgrade in the sandbox store.
> 
> The reason I didn't realize this was the issue is because the free domain notice was coming up for me (probably since the App launches with the regular store, and loads the state for the site), and the purchase was succeeding because I had credits to spend in the sandbox store. Weirdly enough the free-credits purchase succeeds using your credits, but setting the primary domain doesn't.

Yes I think that's the case. To test this, I did not actually use the store sandbox feature, I hardcoded the cookie in the app delegate (without committing it) because of problems related to what you are saying

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift (hunk 5 above)
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift]: Actually, as discussed, removing this notice

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/DomainSuggestionViewControllerWrapper.swift (hunk 2 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/DomainSuggestionViewControllerWrapper.swift]: I'll re-add the second comment to the current code.

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/DomainSuggestionViewControllerWrapper.swift [hunk 7]
```diff
@@ -0,0 +1,51 @@
+import SwiftUI
+import UIKit
+
+/// Makes RegisterDomainSuggestionsViewController available to SwiftUI
+final class DomainSuggestionViewControllerWrapper: UIViewControllerRepresentable {
+
+    private let blog: Blog
+
+    private weak var presentingController: RegisterDomainSuggestionsViewController?
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/DomainSuggestionViewControllerWrapper.swift]: Maybe we should rename this to `presentedViewController`.

I'm using this as my reference for the naming suggestion: https://developer.apple.com/documentation/uikit/uiviewcontroller/1621430-presentingviewcontroller

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift (hunk 4 above)
Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift]: They were actually synonyms and I didn't know.  My apologies.

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/DomainSuggestionViewControllerWrapper.swift (hunk 7 above)
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/DomainSuggestionViewControllerWrapper.swift]: I don't think we should, or at least not that way: this vc is not "presented"

**Review by diegoreymendez — APPROVED** [review]

**Comment by startuptester** [comment]
iPhone 12 Pro, 14.7.1, WP Internal 
I was able to complete all the steps above. 
I sent some minor nitpicks about some of the text to @kyleaparker already. 

One thing I ran into was special characters in the company field made the registration not initiate/fail. (Other fields had validation)

https://user-images.githubusercontent.com/66104034/132590288-c6562358-9338-4a96-bcf1-b7d1ce9a6c9b.MP4

**Pull request merged**